smithy-model/src/main/java/software/amazon/smithy/model/loader/IdlModelLoader.java

```diff
@@ -0,0 +1,879 @@
+        visitor.onShape(builder.id(id));
+        expectNewline();
+    }
+
+    private void parseMap() {
+        // map Foo { key: Bar, value: Baz }
+        SourceLocation sourceLocation = currentLocation();
+        ShapeId id = parseShapeName();
+        MapShape.Builder builder = MapShape.builder()
+                .id(id)
+                .source(sourceLocation);
+
+        parseStructuredContents("map", id, MAP_KEYS);
+        visitor.onShape(builder);
+        expectNewline();
+    }
+
+    private void parseApply() {
+        requireNamespaceOrThrow();
+        // apply <ShapeName> @<trait>\n
+        String name = expect(UNQUOTED).lexeme;
+        ShapeId id = ShapeId.fromOptionalNamespace(visitor.getNamespace(), name);
+        Token token = expect(ANNOTATION);
+        Pair<String, Node> trait = parseTraitValue(token, TraitValueType.APPLY);
+        visitor.onTrait(id, trait.getLeft(), trait.getRight());
+        expectNewline();
+    }
+
+    /**
+     * @return Returns true if the parser is at the last token.
+     */
+    private boolean eof() {
+        return !lexer.hasNext();
+    }
+
+    /**
+     * @return Gets the last valid token.
+     */
+    private Token current() {
+        return Objects.requireNonNull(current, "Call to next must occur before accessing current");
+    }
+
+    /**
+     * @return Get the current source location.
+     */
+    private SourceLocation currentLocation() {
+        return current.getSourceLocation();
+    }
+
+    /**
+     * Expects that there is a next token and returns it.
+     *
+     * @return Returns the next token.
+     * @throws ModelSyntaxException if there is no next token.
+     */
+    private Token next() {
+        if (!lexer.hasNext()) {
+            throw syntax("Unexpected EOF", current != null ? current.span : 0);
+        }
+
+        current = lexer.next();
+        return current;
+    }
+
+    /**
+     * Expects that there is a next token and that it is one of the
+     * given token types.
+     *
+     * @param tokens Token types to expect.
+     * @return Returns the next token.
+     * @throws ModelSyntaxException if the next token is invalid.
+     */
+    private Token expect(TokenType... tokens) {
+        Token next = next();
+        for (TokenType type : tokens) {
+            if (type == next.type) {
+                return next;
+            }
+        }
+
+        if (next.type != ERROR) {
+            throw syntax(tokens.length == 1
+                         ? "Expected " + tokens[0]
+                         : "Expected one of " + Arrays.toString(tokens));
+        } else if (next.lexeme.contains("'") || next.lexeme.contains("\"")) {
+            throw syntax("Unexpected syntax. Perhaps an unclosed quote?");
+        } else if (next.errorMessage == null) {
+            throw syntax("Unexpected syntax");
+        } else {
+            throw syntax(next.errorMessage);
+        }
+    }
+
+    private void parseService() {
```

Comment:
Nit: Reads a little oddly that these parse functions are between some of the token/lexer handling.